File: packages/go/protocol/flow.go
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 DeRec Alliance. All rights reserved.

package protocol

import (
	"encoding/json"
	"errors"
	"fmt"
	"strconv"

	"github.com/derecalliance/lib-derec/packages/go/internal/native"
)

// FlowKind selects which flow Start dispatches. Values match the numeric
// constants in library/src/interop/ffi/protocol/flow.rs.
type FlowKind uint32

const (
	FlowKindPairing           FlowKind = 0
	FlowKindDiscovery         FlowKind = 1
	FlowKindProtectSecret     FlowKind = 2
	FlowKindVerifyShares      FlowKind = 3
	FlowKindRecoverSecret     FlowKind = 4
	FlowKindUnpair            FlowKind = 5
	FlowKindUpdateChannelInfo FlowKind = 6
	// FlowKindReplicaDiscovery asks the replica group whether this device is behind
	// and catches up if it is. Replica-only, and takes no params.
	FlowKindReplicaDiscovery FlowKind = 7
	// FlowKindUnpairReplica removes a member from the replica group.
	// Replica-only: naming this device is a voluntary departure, naming
	// another is an eviction.
	FlowKindUnpairReplica FlowKind = 8
)

// targetKind discriminates Target's three wire shapes. The zero value
// (targetAll) is what the Target zero value carries, so an unconstructed
// Target{} is equivalent to TargetAll() — matching parse_target's
// #[serde(default)] None-means-All behavior.
type targetKind int

const (
	targetAll targetKind = iota
	targetOne
	targetMany
)

// Target selects which channels a fan-out flow (Discovery, VerifyShares,
// UpdateChannelInfo) addresses. The zero value is equivalent to TargetAll.
// Construct via TargetAll, TargetOne, or TargetMany.
type Target struct {
	kind targetKind
	ids  []uint64
}

// TargetAll addresses every paired channel.
func TargetAll() Target { return Target{kind: targetAll} }

// TargetOne addresses a single channel.
func TargetOne(channelID uint64) Target {
	return Target{kind: targetOne, ids: []uint64{channelID}}
}

// TargetMany addresses an explicit set of channels. Note this is distinct
// from TargetAll even when called with zero ids — Rust's parse_target
// resolves an empty array to zero channels, not "every channel".
func TargetMany(channelIDs ...uint64) Target {
	return Target{kind: targetMany, ids: channelIDs}
}

// MarshalJSON implements json.Marshaler, producing the wire shape
// parse_target in library/src/interop/ffi/protocol/flow.rs expects: null for
// TargetAll, a decimal-string for TargetOne, an array of decimal-strings
// for TargetMany.
func (t Target) MarshalJSON() ([]byte, error) {
	switch t.kind {
	case targetOne:
		return json.Marshal(strconv.FormatUint(t.ids[0], 10))
	case targetMany:
		strs := make([]string, len(t.ids))
		for i, id := range t.ids {
			strs[i] = strconv.FormatUint(id, 10)
		}
		return json.Marshal(strs)
	default:
		return []byte("null"), nil
	}
}

// PairingParams are the parameters for FlowKindPairing.
type PairingParams struct {
	// Kind is the local party's role in the handshake. Matches
	// derec_proto::SenderKind's numeric values — the same ones the
	// SenderKind* constants in this package hold (0=Owner, 1=Helper,
	// 3=ReplicaSource, 4=ReplicaDestination), e.g.
	// int32(SenderKindHelper).
	Kind int32
	// Contact is the prost-encoded ContactMessage received out-of-band
	// (the peer's CreatedContact.ContactBytes from CreateContact).
	Contact []byte
	// PeerCommunicationInfo is this node's communication_info to publish
	// to the peer during the handshake. Optional.
	PeerCommunicationInfo map[string]string
}

type pairingParamsWire struct {
	Kind                  int32                `json:"kind"`
	Contact               native.JSONByteArray `json:"contact"`
	PeerCommunicationInfo map[string]string    `json:"peer_communication_info,omitempty"`
}

// DiscoveryParams are the parameters for FlowKindDiscovery.
type DiscoveryParams struct {
	Target Target
}

type discoveryParamsWire struct {
	Target Target `json:"target"`
}

// ProtectSecretParams are the parameters for FlowKindProtectSecret. The
// target set is the protocol's full roster of paired Owner->Helper and
// Source->ReplicaDestination channels; it is not carried on the params.
type ProtectSecretParams struct {
	Secrets []UserSecret
	// Description is optional; nil omits it (matches the Rust side's
	// Option<String> None), a non-nil pointer — including one to an
	// empty string — sets it explicitly.
	Description *string
}

type protectSecretParamsWire struct {
	Secrets     []UserSecret `json:"secrets"`
	Description *string      `json:"description,omitempty"`
}

// VerifySharesParams are the parameters for FlowKindVerifyShares.
type VerifySharesParams struct {
	SecretID uint64
	Version  uint32
	Target   Target
}

type verifySharesParamsWire struct {
	SecretID string `json:"secret_id"`
	Version  uint32 `json:"version"`
	Target   Target `json:"target"`
}

// RecoverSecretParams are the parameters for FlowKindRecoverSecret.
type RecoverSecretParams struct {
	SecretID uint64
	Version  uint32
}

type recoverSecretParamsWire struct {
	SecretID string `json:"secret_id"`
	Version  uint32 `json:"version"`
}

// UnpairParams are the parameters for FlowKindUnpair.
type UnpairParams struct {
	ChannelID uint64
	// Memo is optional; nil omits it.
	Memo *string
}

type unpairParamsWire struct {
	ChannelID string  `json:"channel_id"`
	Memo      *string `json:"memo,omitempty"`
}

// TransportProtocolParam is the transport endpoint carried by
// UpdateChannelInfoParams.
type TransportProtocolParam struct {
	URI      string
	Protocol int32
}

// ReplicaDiscoveryParams carries nothing: the group and this device's own
// version are both read from the stores.
//
// Declared so every FlowKind has a params type, matching dotnet's
// ReplicaDiscoveryParams and the TypeScript `Record<string, never>`. Start
// also accepts nil for this flow.
type ReplicaDiscoveryParams struct{}

// UnpairReplicaParams are the parameters for FlowKindUnpairReplica.
// ReplicaID names the member being removed — this device for a voluntary
// departure, another for an eviction.
type UnpairReplicaParams struct {
	ReplicaID uint64
	Memo      *string
}

type unpairReplicaParamsWire struct {
	ReplicaID string  `json:"replica_id"`
	Memo      *string `json:"memo,omitempty"`
}

// UpdateChannelInfoParams are the parameters for FlowKindUpdateChannelInfo.
type UpdateChannelInfoParams struct {
	Target Target
	// CommunicationInfo replaces the target(s)' view of this node's
	// communication_info. nil leaves it untouched; a non-nil map
	// (including an empty one) sets it, matching the destructive-replace
	// semantics of SetCommunicationInfo.
	CommunicationInfo map[string]string
	// TransportProtocol replaces the target(s)' view of this node's
	// transport endpoint. nil leaves it untouched.
	//
	// Deprecated: superseded by OwnTransports, which carries every endpoint
	// rather than one. Scheduled for removal in v0.0.5. OwnTransports takes
	// precedence when both are set.
	TransportProtocol *TransportProtocolParam
	// OwnTransports replaces the target(s)' view of every endpoint this node
	// serves, in its own preference order. Empty leaves them untouched. The
	// first entry also fills the deprecated singular field so a peer
	// predating the list still learns the new address.
	OwnTransports []TransportProtocolParam
}

type updateChannelInfoParamsWire struct {
	Target            Target                       `json:"target"`
	CommunicationInfo *map[string]string           `json:"communication_info,omitempty"`
	TransportProtocol *transportProtocolParamWire  `json:"transport_protocol,omitempty"`
	OwnTransports     []transportProtocolParamWire `json:"own_transports,omitempty"`
}

type transportProtocolParamWire struct {
	URI      string `json:"uri"`
	Protocol int32  `json:"protocol"`
}

// marshalFlowParams validates params against flowKind and renders it into
// the JSON shape the matching decoder in library/src/interop/ffi/protocol/flow.rs
// expects.
func marshalFlowParams(flowKind FlowKind, params any) ([]byte, error) {
	switch flowKind {
	case FlowKindPairing:
		pp, ok := params.(PairingParams)
		if !ok {
			return nil, fmt.Errorf("protocol: Start: FlowKindPairing requires PairingParams, got %T", params)
		}
		return json.Marshal(pairingParamsWire{
			Kind:                  pp.Kind,
			Contact:               native.JSONByteArray(pp.Contact),
			PeerCommunicationInfo: pp.PeerCommunicationInfo,
		})
	case FlowKindDiscovery:
		dp, ok := params.(DiscoveryParams)
		if !ok {
			return nil, fmt.Errorf("protocol: Start: FlowKindDiscovery requires DiscoveryParams, got %T", params)
		}
		return json.Marshal(discoveryParamsWire{Target: dp.Target})
	case FlowKindProtectSecret:
		psp, ok := params.(ProtectSecretParams)
		if !ok {
			return nil, fmt.Errorf("protocol: Start: FlowKindProtectSecret requires ProtectSecretParams, got %T", params)
		}
		secrets := psp.Secrets
		if secrets == nil {
			// A nil slice with no `omitempty` marshals as JSON null, not
			// `[]` — the Rust decoder's `secrets: Vec<UserSecret>` field
			// isn't Option-wrapped, so it would reject a null. Normalize
			// to a non-nil empty slice to keep marshaling `[]`.
			secrets = []UserSecret{}
		}
		return json.Marshal(protectSecretParamsWire{
			Secrets:     secrets,
			Description: psp.Description,
		})
	case FlowKindVerifyShares:
		vsp, ok := params.(VerifySharesParams)
		if !ok {
			return nil, fmt.Errorf("protocol: Start: FlowKindVerifyShares requires VerifySharesParams, got %T", params)
		}
		return json.Marshal(verifySharesParamsWire{
			SecretID: strconv.FormatUint(vsp.SecretID, 10),
			Version:  vsp.Version,
			Target:   vsp.Target,
		})
	case FlowKindRecoverSecret:
		rsp, ok := params.(RecoverSecretParams)
		if !ok {
			return nil, fmt.Errorf("protocol: Start: FlowKindRecoverSecret requires RecoverSecretParams, got %T", params)
		}
		return json.Marshal(recoverSecretParamsWire{
			SecretID: strconv.FormatUint(rsp.SecretID, 10),
			Version:  rsp.Version,
		})
	case FlowKindUnpair:
		up, ok := params.(UnpairParams)
		if !ok {
			return nil, fmt.Errorf("protocol: Start: FlowKindUnpair requires UnpairParams, got %T", params)
		}
		return json.Marshal(unpairParamsWire{
			ChannelID: strconv.FormatUint(up.ChannelID, 10),
			Memo:      up.Memo,
		})
	case FlowKindUpdateChannelInfo:
		ucip, ok := params.(UpdateChannelInfoParams)
		if !ok {
			return nil, fmt.Errorf("protocol: Start: FlowKindUpdateChannelInfo requires UpdateChannelInfoParams, got %T", params)
		}
		w := updateChannelInfoParamsWire{Target: ucip.Target}
		if ucip.CommunicationInfo != nil {
			w.CommunicationInfo = &ucip.CommunicationInfo
		}
		if len(ucip.OwnTransports) > 0 {
			w.OwnTransports = make([]transportProtocolParamWire, len(ucip.OwnTransports))
			for i, t := range ucip.OwnTransports {
				w.OwnTransports[i] = transportProtocolParamWire{URI: t.URI, Protocol: t.Protocol}
			}
			// The first entry also fills the deprecated singular field.
			w.TransportProtocol = &w.OwnTransports[0]
		} else if ucip.TransportProtocol != nil {
			w.TransportProtocol = &transportProtocolParamWire{
				URI:      ucip.TransportProtocol.URI,
				Protocol: ucip.TransportProtocol.Protocol,
			}
		}
		return json.Marshal(w)
	case FlowKindUnpairReplica:
		rrp, ok := params.(UnpairReplicaParams)
		if !ok {
			return nil, fmt.Errorf("protocol: Start: FlowKindUnpairReplica requires UnpairReplicaParams, got %T", params)
		}
		return json.Marshal(unpairReplicaParamsWire{
			ReplicaID: strconv.FormatUint(rrp.ReplicaID, 10),
			Memo:      rrp.Memo,
		})
	case FlowKindReplicaDiscovery:
		// No parameters: the group and this device's own version both come
		// from the stores. nil is accepted for the same reason the other
		// SDKs make theirs optional.
		if params != nil {
			if _, ok := params.(ReplicaDiscoveryParams); !ok {
				return nil, fmt.Errorf("protocol: Start: FlowKindReplicaDiscovery requires ReplicaDiscoveryParams or nil, got %T", params)
			}
		}
		return json.Marshal(struct{}{})
	default:
		return nil, fmt.Errorf("protocol: Start: unknown FlowKind %d", flowKind)
	}
}

// Start kicks off flowKind with the matching params struct — PairingParams,
// DiscoveryParams, ProtectSecretParams, VerifySharesParams,
// RecoverSecretParams, UnpairParams, UpdateChannelInfoParams,
// ReplicaDiscoveryParams or UnpairReplicaParams. Passing a
// params value that doesn't match flowKind is a returned error, not a
// panic. Returns the per-target *Started / *Failed events describing what
// was dispatched.
func (p *DeRecProtocol) Start(flowKind FlowKind, params any) ([]Event, error) {
	if p.closed {
		return nil, errors.New("protocol: Start: protocol is closed")
	}
	paramsJSON, err := marshalFlowParams(flowKind, params)
	if err != nil {
		return nil, err
	}
	eventsJSON, err := p.instance.Start(uint32(flowKind), paramsJSON)
	if err != nil {
		return nil, err
	}
	return decodeEvents(eventsJSON)
}

// Accept accepts a pending action carried by an ActionRequired event's
// Action field, passed verbatim. Returns the events produced by resuming
// the suspended flow (e.g. PairingCompleted).
func (p *DeRecProtocol) Accept(action []byte) ([]Event, error) {
	if p.closed {
		return nil, errors.New("protocol: Accept: protocol is closed")
	}
	eventsJSON, err := p.instance.Accept(action)
	if err != nil {
		return nil, err
	}
	return decodeEvents(eventsJSON)
}

// Reject rejects a pending action carried by an ActionRequired event's
// Action field, with a status + memo. status matches
// derec_proto::StatusEnum.
func (p *DeRecProtocol) Reject(action []byte, status int32, memo string) error {
	if p.closed {
		return errors.New("protocol: Reject: protocol is closed")
	}
	return p.instance.Reject(action, status, memo)
}

// Restore rebuilds this protocol's secret_id namespace from a recovered
// Secret — the same typed snapshot carried by SecretRecoveredEvent.Secret
// (and ReplicaSecretReceivedEvent.Secret); pass it verbatim along with the
// version it was recovered at.
func (p *DeRecProtocol) Restore(secret Secret, version uint32) ([]Event, error) {
	if p.closed {
		return nil, errors.New("protocol: Restore: protocol is closed")
	}
	paramsJSON, err := json.Marshal(restoreParamsWire{
		Version:         version,
		RecoveredSecret: secret,
	})
	if err != nil {
		return nil, fmt.Errorf("protocol: Restore: marshal params: %w", err)
	}
	eventsJSON, err := p.instance.Restore(paramsJSON)
	if err != nil {
		return nil, err
	}
	return decodeEvents(eventsJSON)
}

// restoreParamsWire carries the version alongside Secret's own MarshalJSON
// output (see events.go) — no separate wire mirror is needed since Secret
// already encodes to the byte-array-as-number-array shape Restore's FFI
// params require.
type restoreParamsWire struct {
	Version         uint32 `json:"version"`
	RecoveredSecret Secret `json:"recovered_secret"`
}

File: library/src/interop/ffi/protocol/flow.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 DeRec Alliance. All rights reserved.

//! JSON decoder for [`DeRecFlow`] params passed via the FFI
//! `derec_protocol_start` entry point.
//!
//! The dotnet caller serializes the flow params as a JSON object with
//! shape matching the per-variant struct below. The Rust side decodes
//! and constructs the matching `DeRecFlow` value.
//!
//! Targeting convention (mirrors the WASM bridge): the `target` JSON
//! field is either `null` (= [`Target::All`]), a single u64 string
//! (= [`Target::Single`]), or an array of u64 strings
//! (= [`Target::Many`]). u64s travel as decimal strings to dodge
//! JavaScript's 53-bit integer ceiling — dotnet just serialises them
//! the same way.

use std::collections::HashMap;

use derec_proto::{ContactMessage, SenderKind, TransportProtocol};
use prost::Message as _;
use serde::Deserialize;
use serde_json::Value;

use crate::protocol::DeRecFlow;
use crate::protocol::types::{Target, UserSecret};
use crate::types::ChannelId;

/// Numeric flow-kind identifiers — must match the dotnet `FlowKind` enum.
pub const FLOW_KIND_PAIRING: u32 = 0;
pub const FLOW_KIND_DISCOVERY: u32 = 1;
pub const FLOW_KIND_PROTECT_SECRET: u32 = 2;
pub const FLOW_KIND_VERIFY_SHARES: u32 = 3;
pub const FLOW_KIND_RECOVER_SECRET: u32 = 4;
pub const FLOW_KIND_UNPAIR: u32 = 5;
pub const FLOW_KIND_UPDATE_CHANNEL_INFO: u32 = 6;
/// Replica catch-up. Takes no parameters: the group and this device's own
/// version are both read from the stores.
pub const FLOW_KIND_REPLICA_DISCOVERY: u32 = 7;
/// Remove a member from the replica group. Params:
/// `{ "replica_id": "<decimal>", "memo": "<optional>" }`.
pub const FLOW_KIND_UNPAIR_REPLICA: u32 = 8;

/// Top-level dispatcher — picks the right decoder based on `flow_kind`.
pub fn parse_flow(flow_kind: u32, params_json: &[u8]) -> Result<DeRecFlow, String> {
    match flow_kind {
        FLOW_KIND_PAIRING => parse_pairing_flow(params_json),
        FLOW_KIND_DISCOVERY => parse_discovery_flow(params_json),
        FLOW_KIND_PROTECT_SECRET => parse_protect_secret_flow(params_json),
        FLOW_KIND_VERIFY_SHARES => parse_verify_shares_flow(params_json),
        FLOW_KIND_RECOVER_SECRET => parse_recover_secret_flow(params_json),
        FLOW_KIND_UNPAIR => parse_unpair_flow(params_json),
        FLOW_KIND_UPDATE_CHANNEL_INFO => parse_update_channel_info_flow(params_json),
        FLOW_KIND_REPLICA_DISCOVERY => Ok(DeRecFlow::ReplicaDiscovery),
        FLOW_KIND_UNPAIR_REPLICA => parse_unpair_replica_flow(params_json),
        other => Err(format!("unknown FlowKind: {other}")),
    }
}

fn parse_pairing_flow(params_json: &[u8]) -> Result<DeRecFlow, String> {
    let raw: PairingParamsJson = serde_json::from_slice(params_json)
        .map_err(|e| format!("invalid PairingParams JSON: {e}"))?;
    let kind = sender_kind_from_i32(raw.kind)?;
    let contact = ContactMessage::decode(raw.contact.as_slice())
        .map_err(|e| format!("invalid ContactMessage proto bytes: {e}"))?;
    let peer_communication_info = raw.peer_communication_info.unwrap_or_default();
    Ok(DeRecFlow::Pairing {
        kind,
        contact,
        peer_communication_info,
    })
}

fn parse_discovery_flow(params_json: &[u8]) -> Result<DeRecFlow, String> {
    let raw: DiscoveryParamsJson = serde_json::from_slice(params_json)
        .map_err(|e| format!("invalid DiscoveryParams JSON: {e}"))?;
    Ok(DeRecFlow::Discovery {
        target: parse_target(raw.target)?,
    })
}

fn parse_protect_secret_flow(params_json: &[u8]) -> Result<DeRecFlow, String> {
    let raw: ProtectSecretParamsJson = serde_json::from_slice(params_json)
        .map_err(|e| format!("invalid ProtectSecretParams JSON: {e}"))?;
    let secrets: Vec<UserSecret> = raw
        .secrets
        .into_iter()
        .map(|s| UserSecret {
            id: s.id,
            name: s.name,
            data: s.data,
        })
        .collect();
    Ok(DeRecFlow::ProtectSecret {
        secrets,
        description: raw.description,
    })
}

fn parse_verify_shares_flow(params_json: &[u8]) -> Result<DeRecFlow, String> {
    let raw: VerifySharesParamsJson = serde_json::from_slice(params_json)
        .map_err(|e| format!("invalid VerifySharesParams JSON: {e}"))?;
    Ok(DeRecFlow::VerifyShares {
        secret_id: parse_u64_string(&raw.secret_id)?,
        version: raw.version,
        target: parse_target(raw.target)?,
    })
}

fn parse_recover_secret_flow(params_json: &[u8]) -> Result<DeRecFlow, String> {
    let raw: RecoverSecretParamsJson = serde_json::from_slice(params_json)
        .map_err(|e| format!("invalid RecoverSecretParams JSON: {e}"))?;
    Ok(DeRecFlow::RecoverSecret {
        secret_id: parse_u64_string(&raw.secret_id)?,
        version: raw.version,
    })
}

fn parse_unpair_flow(params_json: &[u8]) -> Result<DeRecFlow, String> {
    let raw: UnpairParamsJson = serde_json::from_slice(params_json)
        .map_err(|e| format!("invalid UnpairParams JSON: {e}"))?;
    Ok(DeRecFlow::Unpair {
        channel_id: ChannelId(parse_u64_string(&raw.channel_id)?),
        memo: raw.memo,
    })
}

fn parse_update_channel_info_flow(params_json: &[u8]) -> Result<DeRecFlow, String> {
    let raw: UpdateChannelInfoParamsJson = serde_json::from_slice(params_json)
        .map_err(|e| format!("invalid UpdateChannelInfoParams JSON: {e}"))?;
    // The list wins when present; the singular field is the compatibility
    // spelling an SDK predating it still sends.
    let own_transports: Vec<TransportProtocol> = if raw.own_transports.is_empty() {
        raw.transport_protocol
            .into_iter()
            .map(|t| TransportProtocol {
                uri: t.uri,
                protocol: t.protocol,
            })
            .collect()
    } else {
        raw.own_transports
            .into_iter()
            .map(|t| TransportProtocol {
                uri: t.uri,
                protocol: t.protocol,
            })
            .collect()
    };
    Ok(DeRecFlow::UpdateChannelInfo {
        target: parse_target(raw.target)?,
        communication_info: raw.communication_info,
        own_transports,
    })
}

fn sender_kind_from_i32(kind: i32) -> Result<SenderKind, String> {
    match kind {
        0 => Ok(SenderKind::Owner),
        1 => Ok(SenderKind::Helper),
        3 => Ok(SenderKind::ReplicaSource),
        4 => Ok(SenderKind::ReplicaDestination),
        other => Err(format!("invalid SenderKind: {other}")),
    }
}

fn parse_u64_string(s: &str) -> Result<u64, String> {
    s.parse::<u64>()
        .map_err(|e| format!("expected decimal u64 string, got {s:?}: {e}"))
}

fn parse_target(value: Option<Value>) -> Result<Target, String> {
    match value {
        None | Some(Value::Null) => Ok(Target::All),
        Some(Value::String(s)) => Ok(Target::Single(ChannelId(parse_u64_string(&s)?))),
        Some(Value::Number(n)) => {
            let id = n
                .as_u64()
                .ok_or_else(|| format!("Target number {n} is not a u64"))?;
            Ok(Target::Single(ChannelId(id)))
        }
        Some(Value::Array(arr)) => {
            let mut ids = Vec::with_capacity(arr.len());
            for item in arr {
                let id = match item {
                    Value::String(s) => parse_u64_string(&s)?,
                    Value::Number(n) => n
                        .as_u64()
                        .ok_or_else(|| format!("Target array item {n} is not a u64"))?,
                    other => return Err(format!("Target array item must be u64, got {other:?}")),
                };
                ids.push(ChannelId(id));
            }
            Ok(Target::Many(ids))
        }
        Some(other) => Err(format!(
            "Target must be null, u64 string, or u64 array — got {other:?}"
        )),
    }
}

#[derive(Deserialize)]
struct PairingParamsJson {
    kind: i32,
    contact: Vec<u8>,
    #[serde(default)]
    peer_communication_info: Option<HashMap<String, String>>,
}

#[derive(Deserialize)]
struct DiscoveryParamsJson {
    #[serde(default)]
    target: Option<Value>,
}

#[derive(Deserialize)]
struct ProtectSecretParamsJson {
    secrets: Vec<UserSecretJson>,
    #[serde(default)]
    description: Option<String>,
}

#[derive(Deserialize)]
struct UserSecretJson {
    id: Vec<u8>,
    name: String,
    data: Vec<u8>,
}

#[derive(Deserialize)]
struct VerifySharesParamsJson {
    secret_id: String,
    version: u32,
    #[serde(default)]
    target: Option<Value>,
}

#[derive(Deserialize)]
struct RecoverSecretParamsJson {
    secret_id: String,
    version: u32,
}

#[derive(Deserialize)]
struct UnpairParamsJson {
    channel_id: String,
    #[serde(default)]
    memo: Option<String>,
}

#[derive(Deserialize)]
struct UpdateChannelInfoParamsJson {
    #[serde(default)]
    target: Option<Value>,
    #[serde(default)]
    communication_info: Option<HashMap<String, String>>,
    /// Deprecated: superseded by `own_transports`. Accepted so an SDK
    /// predating the list keeps working; removal is scheduled for v0.0.5.
    #[serde(default)]
    transport_protocol: Option<TransportProtocolJson>,
    /// Every endpoint this device can now be reached on, in preference
    /// order. Takes precedence over the singular field above.
    #[serde(default)]
    own_transports: Vec<TransportProtocolJson>,
}

#[derive(Deserialize)]
struct TransportProtocolJson {
    uri: String,
    protocol: i32,
}

/// `{ "replica_id": "<decimal u64>", "memo": "<optional>" }` — `replica_id` is
/// a decimal string so values above 2^53 survive a JSON round trip through
/// hosts whose numbers are doubles.
fn parse_unpair_replica_flow(params_json: &[u8]) -> Result<DeRecFlow, String> {
    #[derive(serde::Deserialize)]
    struct Params {
        replica_id: String,
        #[serde(default)]
        memo: Option<String>,
    }
    let p: Params = serde_json::from_slice(params_json)
        .map_err(|e| format!("UnpairReplica params JSON: {e}"))?;
    let replica_id = p
        .replica_id
        .parse::<u64>()
        .map_err(|e| format!("replica_id must be a decimal u64: {e}"))?;
    Ok(DeRecFlow::UnpairReplica {
        replica_id,
        memo: p.memo,
    })
}
